lib/chef/provisioning/convergence_strategy/install_cached.rb

```diff
@@ -42,14 +42,26 @@ def initialize(convergence_options, config)
       def setup_convergence(action_handler, machine)
         super
 
-        # Install chef-client.  TODO check and update version if not latest / not desired
-        if machine.execute_always('chef-client -v').exitstatus != 0
-          platform, platform_version, machine_architecture = machine.detect_os(action_handler)
-          package_file = download_package_for_platform(action_handler, machine, platform, platform_version, machine_architecture)
-          remote_package_file = "#{@tmp_dir}/#{File.basename(package_file)}"
-          machine.upload_file(action_handler, package_file, remote_package_file)
-          install_package(action_handler, machine, remote_package_file)
+        # Check for existing chef client.
+        version = machine.execute_always('chef-client -v')
+
+        # Don't do install/upgrade if a chef client exists and
+        # no chef version is defined by user configs or
+        # the chef client's version already matches user config
+        if version.exitstatus == 0
+          if !chef_version
+            return
+          elsif version.stdout.strip =~ /Chef: #{chef_version}([^0-9]|$)/
```

Comment:
The thing we need to make sure of is that we don't print green text (and mark the resource updated) of the script decides no upgrade needs to be performed. That will take some doing, but I'd be happy to help next week. Meantime, do you feel we should take the patch and file a bug for that case, try to fix first, or wait?